### ISSUE TITLE
How to split by sentence

### ISSUE THREAD

**Issue by nachocab** [kach/nearley#365]
I've gone through all the nearley documentation and I've just started writing my first grammars, but I'm running into some confusion. I don't know if this the right place to post my question, so please let me know if I should move it somewhere else.

Using this input `bb.      c. ddd.    ` (characters followed by a period and 1+ spaces)

I'm trying to get this output `[[ 'bb.', 'c.', 'ddd.']]` (an array of sentences, including the period)

But this is what I'm getting:

    [[ [ [ [ 'bb', '.' ], 'c', '.' ], 'ddd', '.' ] ]]

This is the grammar I'm using

    @{%
    const handle2 = (d) => {
      const x = d[0].concat(d[1])
      const y = d[2]
      return [x, y]
    }

    const handle = (d) => {
      const x = d[0].concat(d[1])
      return x
    }
    %}

    main -> sentences
    sentences -> 
        sentences terminator word {% handle2 %}
      | sentences terminator {% handle %}
      | word 
    word -> [a-z]:+ {% d => d[0].join('') %}
    terminator -> 
        [.] " ":* {% d => d[0] %}

I think my main problem is that I get lost trying to follow what each postprocessor gets and what it should return. I'm especially confused about left-recursion. I don't know if a better pattern would be to use `sentences -> word (terminator word:+):* terminator` or something like that.

Any help or pointers would be very appreciated.

**Comment by nachocab** [comment]
Well, it took me 5:30 hours, but I think I understand BNF and postprocessors much better now. This is what I was looking for:

```
sentences -> sentence 
  | sentences _ sentence {% d => d[0].concat(d[2])  %}
sentence -> chars terminator {% (d) => d[0].concat(d[1]).join('') %}
chars -> [a-z]:+ {% d => d[0] %}
terminator -> "."
_ -> " ":+ 
```
I think what was tripping me up was not having a clear idea of the arguments that each rule was using to call the postprocessor function. 

It might be a good idea to add something like this to the [How to grammar good](https://nearley.js.org/docs/how-to-grammar-good) section:

### Postprocessors
If you're not sure about what postprocessor you should use, write your grammar without them. Then, start adding them from the bottom. Use `nearley-test` to check your input string to see the effect of each new postprocessor. Remember that the number of elements that the postprocessor function will receive will depend on the number of elements in your rule. For example, if you want to concat `B` and `E` in the rule below, you should write a postprocessor that targets elements 2 and 5:

```
S -> A _ B _ "," E  {% d => d[2].concat(d[5]) %}
```
Also, remember that a processor only affects one side of rule with an alternative. This will not work:

```
# The postprocessor is only called when `A` is matched, so it will only have 1 element
S -> A _ B _ "," E  | A  {% d => d[2].concat(d[5]) %}
```

If you want both sides of the alternative to use the same processor, use capturing groups:

```
S -> (A _ B _ "," E  | A _ C _ "," F)   {% d => d[2].concat(d[5]) %}
```

**Comment by nachocab** [comment]
Also, I found [this BNF guide](http://marvin.cs.uidaho.edu/Teaching/CS445/grammar.pdf) to be a great resource. It might be worth adding to the Further Reading section in [Getting Started](https://nearley.js.org/docs/getting-started)

**Comment by matthew-dean** [comment]
Yeah, the docs could use a lot more examples. It's pretty damn Spartan.
